plugin/src/main/java/org/openrewrite/gradle/isolated/DefaultProjectParser.java

```diff
@@ -705,17 +701,20 @@ public List<SourceFile> parse(Project subproject, Set<Path> alreadyParsed, Execu
                     Duration parseDuration = Duration.between(start, Instant.now());
                     logger.info("Finished parsing Java sources from {}/{} in {} ({} per source)",
                             subproject.getName(), sourceSet.getName(), prettyPrint(parseDuration), prettyPrint(parseDuration.dividedBy(javaPaths.size())));
-                    sourceFiles.addAll(map(autodetectStyle(cus), addProvenance(projectProvenance, null)));
-                    sourceSetProvenance = jp.getSourceSet(ctx); // Hold onto provenance to apply it to resource files
+                    List<JavaType.FullyQualified> classpath = sourceSetProvenance.getClasspath();
+                    for (J.CompilationUnit cu : cus) {
+                        for (JavaType type : cu.getTypesInUse().getTypesInUse()) {
+                            if (type instanceof JavaType.FullyQualified) {
+                                classpath.add((JavaType.FullyQualified) type);
+                            }
+                        }
+                    }
+                    JavaSourceSet javaSourceSetProvenance = sourceSetProvenance.withClasspath(classpath);
+                    sourceFiles.addAll(map(autodetectStyle(cus), addProvenance(projectProvenance, javaSourceSetProvenance)));
```

Comment:
All java/kotlin/groovy types are visible to eachother within the source set (assuming you're using java + groovy OR java + kotlin, all three mixed together makes for difficult build configurations in my experience). 

What if we don't put any provenance markers on until all three sources of types have been parsed, with their types added to a single set (to avoid duplicate entries), then all sources are given the same instance of the same `JavaSourceSet` all at once? Feels like that would be most efficient / de-duplicated.